src/gpu/kmeans/kmeans_init.cu

```diff
@@ -0,0 +1,439 @@
+
+  distance_pairs = centroids_distance_op(_centroids, _centroids);
+
+  // get the closest x_j for each x_i in centroids.
+  KmMatrix<int> min_indices = SelfArgMinOp<T>().argmin(distance_pairs);
+
+  // use historgram to get counting for weights
+  KmMatrix<int> weights (1, _centroids.rows());
+
+  size_t temp_storage_bytes = 0;
+  void *d_temp_storage = NULL;
+
+  // determine the temp_storage_bytes
+  safe_cuda(cub::DeviceHistogram::HistogramEven(
+      d_temp_storage, temp_storage_bytes,
+      min_indices.dev_ptr(),
+      weights.dev_ptr(),
+      min_indices.rows() + 1,
+      (T)0.0,
+      (T)min_indices.rows(),
+      (int)_centroids.rows()));
+
+  safe_cuda(cudaMalloc((void**)&d_temp_storage, temp_storage_bytes));
+  safe_cuda(cub::DeviceHistogram::HistogramEven(
+      d_temp_storage, temp_storage_bytes,
+      min_indices.dev_ptr(),    // d_samples
+      weights.dev_ptr(),        // d_histogram
+      min_indices.rows() + 1,   // num_levels
+      (T)0.0,                   // lower_level
+      (T)min_indices.rows(),    // upper_level
+      (int)_centroids.rows())); // num_samples
+  safe_cuda(cudaFree(d_temp_storage));
+
+  // Sort the indices by weights in ascending order, then use those at front
+  // as result.
+  thrust::sort_by_key(thrust::device,
+                      weights.dev_ptr(),
+                      weights.dev_ptr() + weights.size(),
+                      min_indices.dev_ptr(),
+                      thrust::greater<int>());
+
+  int * min_indices_ptr = min_indices.dev_ptr();
+
+  KmMatrix<T> new_centroids (_k, _centroids.cols());
+  T * new_centroids_ptr = new_centroids.dev_ptr();
+  int cols = _centroids.cols();
+
+  T * old_centroids_ptr = _centroids.dev_ptr();
+
+  auto k_iter = thrust::make_counting_iterator(0);
+  thrust::for_each(
+      thrust::device,
+      k_iter, k_iter + _k,
+      [=] __device__ (int idx) {
+        size_t index = min_indices_ptr[idx];
+
+        size_t in_begin = index * cols;
+        size_t in_end = (index + 1) * cols;
+
+        size_t res_begin = idx * cols;
+
+        for (size_t i = in_begin, j = res_begin; i < in_end; ++i, ++j) {
+          new_centroids_ptr[j] = old_centroids_ptr[i];
+        }
+      });
+
+  return new_centroids;
+}
+
+}  // namespace detail
+
+
+/* ============ KmeansRandomInit Class member functions ============ */
+
+template <typename T>
+KmMatrix<T> KmeansRandomInit<T>::operator()(KmMatrix<T>& _data, size_t _k) {
+
+  KmMatrix<T> dist = generator_impl_->generate(_k);
+  MulOp<T>().mul(dist, dist, (T)_data.rows() - 1);
+
+  dist = dist.reshape(dist.cols(), dist.rows());
+
+  KmMatrix<T> centroids = _data.rows(dist);
+
+  return centroids;
+}
+
+
+/* ============== KmeansLlInit Class member functions ============== */
+
+// Although the paper suggested calculating the probability independently,
+// but due to zero distance between a point and itself, the already selected
+// points will have very low probability.
```

Comment:
Just to confirm it should be impossible to select the same starting point twice correct?

If the same starting point is selected twice the algorithm will not work correctly.

Reply:
It's possible. The odds are low when dataset isn't the toy from test, but possible. And that's one of the tricky draw backs of kmeans||.